src/transformStateWithClones.js

```diff
@@ -5,9 +5,47 @@
  * @param {Object[]} actions
  *
  * @return {Object[]}
+ *
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  /**  write code here
+  const stateHistory = [];
+  const copyState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(copyState, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      for (const element of action.keysToRemove) {
+        delete copyState[element];
+      }
+    } else if (action.type === 'clear') {
+      for (const key of Object.keys(copyState)) {
+        delete copyState[key];
+      }
+    }
+  */
+
+  const stateHistory = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(stateCopy, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      for (const element of action.keysToRemove) {
+        delete stateCopy[element];
+      }
+    } else if (action.type === 'clear') {
+      for (const key of Object.keys(stateCopy)) {
+        delete stateCopy[key];
+      }
+    }
+
+    stateHistory.push(Object.assign({}, stateCopy));
```

Comment:
The action of pushing the state copy into the stateHistory array is currently outside of the if-else or switch block. This should be inside each case to ensure that you're working with a fresh copy of the state for each action, which is a requirement to avoid mutating the original state object.

Reply:
ok